### DISCUSSION TITLE
Implementation of k-epsilon turbulence model(channel flow test cas).

### DISCUSSION THREAD

**Discussion by SoorajBharan** [exadg/exadg#714] [General]
I have been trying to implement the k-epsilon turbulence model in ExaDG. 
For preserving the positivity of k and epsilon natural log of the k and epsilon are taken as the transport variable. The following are the transport equation and the equation for eddy viscosity:
$\mathcal{K} =  ln(k); \mathcal{E} = ln(\epsilon) $
![k_epsilon_transport_equations](https://github.com/user-attachments/assets/1f334169-9a84-45f5-baac-05e002639123)

About the solver:-
I used the following solvers : 1) incompressible_navier_stokes, 2) convection_diffusion, and 3) incompressible_flow_with_transport.
I modified the convection_diffusion solver to have 4 additional terms on the rhs
1)Production term
2)Dissipation term
3) Cross derivative term to account for variable eddy viscosity
4) Cross derivative term to account for using ln of k and epsilon
The implementation of the production and dissipation terms are added as a body force in the rhs_operator.h
The cross derivative terms are computed in the diffusive_operator.h as volume integrals.
I have atteched the modified files here.
[operators.zip](https://github.com/user-attachments/files/18124120/operators.zip)

I am communicating the values of k an epsilon between the incompressible_navier_stokes solver and convection_diffusion solver to update the eddy viscosity.
The Navier-Stokes equations are solver using Pressure Correction method and the backwards difference formulation is used for the transport equations. I used the CG solver for solving the system of equations for the transport equations.

About the test case:-
I decided to test the implementation for the flow through channel at $Re_{\tau}$ = 180; $Re_{\delta} $= 3300
$\delta = 0.5$; $\nu = 1.0e-2$; 
Length of the channel = $4\pi \delta$; 
U is calculated based on $Re_{\delta}$ and parabolic velocity profile is given at the center;
Boundary conditions:
![channel_flow_boundary_condition](https://github.com/user-attachments/assets/54b2b518-4676-4554-8126-b1c5a8a30993)
refinement levels used: 6,7,8
polynomial order used for U:2,3
polynomial order used for p, $\mathcal{K}, \mathcal{E}$ : 1,2

I was able to run the solver without any additional terms than the advection and diffusion terms. However the solver shows divergence when other terms are added with the exception of dissipation term.

Scatter plot of production term after the first iteration:
![tke_production_scatter](https://github.com/user-attachments/assets/1248db62-7fa4-425d-a515-202bb3801ff5)
Scatter plot of Cross derivative for variable eddy viscosity after the first iteration:
![limiter_scatter](https://github.com/user-attachments/assets/51e6810b-52a5-44ce-80c0-403cce68cc39)
Scatter plot of Cross derivative for positivity limiter after the first iteration:
![variable_viscosity_scatter](https://github.com/user-attachments/assets/1a4c6bf8-6b3c-43b3-9634-fd1dba6a7ad7)
In the above figures I have kept the length of channel small to save computational time.

I would greatly appreciate it if you could provide any insights or suggestions on how to proceed. If there is something in my implementation that might be causing the divergence, or if there is a better approach, your guidance would be invaluable.

**Comment by richardschu** [comment]
Hi @SoorajBharan, we are happy to see that you are developing your tools in ExaDG!
Thanks also for the short intro, I have some questions:
1) $\mathcal{K}$ and $\mathcal{\Epsilon}$ are added as scalar variables and you solve for them to update the apparent viscosity, right?
2) the viscosity is the updated in the Navier-Stokes solver,do you use the generalizedNewtonian framework for this?
3) did you also change the projection solvers? ('BDFDualSplitting' and 'BDFPressureCorrection'?) ... I am currently working on some solvers to consider variable viscosity as is the case for you, but I have no clue, if the viscous terms and consistency of the projection solvers are in your case as important as they are in my scenario (which is low to average Reynolds number flow).

And it would be great if you could make a draft pull request so that we can check out the code directly on github.

From the plots I cannot really say anything tbh apart from: "something is happening at the inlet".

Speaking from experience it would make sense if you tried to isolate the problems and look at smaller and smaller parts of the code. For example: do you even get a single time step solved? how does it look like? can you prescribe the solutions to the $\mathcal{K}$ and $\mathcal{\Epsilon}$ and the Navier-Stokes solver runs fine with the adapted viscosity? can you just passively run the solvers for the $\mathcal{K}$ and $\mathcal{\Epsilon}$ and not put any info back in the Navier-Stokes solver? Then the problem might only be in the added scalar solvers. Maybe start from an artificial problem where you can isolate effects? 

Please provide more info and a draft PR for us to have a look!

**Reply by richardschu** [reply]
oh and could you provide the above equations in non-component version? Then it would be easier to understand for me :)

**Reply by SoorajBharan** [reply]
Hi @richardschu , Thank you for responding to my question.

First I will answer your questions:
1) $\mathcal{K}[\ln(k)]$ and $\mathcal{E}[ln(\epsilon)]$ are obtained by solving scalar transport equation using the convection-diffusion solver. Currently I am passing these value to incompressible_navier_stokes solver to calculate the apparent viscosity.

2) I modified  the TurbulenceModel(derived from ViscosityModelBase) to calculate the eddy viscosity and add it to the kinematic viscosity. I have also created the same classes in the convection_diffusion solver so that the eddy viscosity is accessible to the objects in the namespace ConvDiff. I use the TransportModel class to store the values of $\mathcal{K}$ and $\mathcal{E}$ as VectorType arrays.

3) I have only used BDFPressureCorrection scheme. Since ExaDG already has a turbulence model with variable viscosity I thought it would work.

About the cuurent problem:
Even with a small gradient in velocity or other transport variables($\mathcal{K}$ and $\mathcal{E}$) the equations doesn't solves for even iteration. But when I only have the temporal , advection, diffusion and dissipation term, with solver converges with uniform initial conditions all over the domain. 

I will definitely prepare a draft pull for you to review.

Thanks again for your help. 

![k_epsilon_transport_equations_vector_notation](https://github.com/user-attachments/assets/b6c8ea3c-dc54-4b66-a06c-73f3986255cc)

**Reply by richardschu** [reply]
Ok so one thing I can already say is that the pressure-correction solver in its current version is not fully consistent. Especially on the boundaries, we expect increased errors, but I am not sure this is causing your problem. What if you us the `BDFCoupledScheme` just to rule out such issues? You can still use the variable viscosity to store the fields to access them during element integration and the convection-diffusion solvers to compute the scalar components!
I unfortunately do not understand what you mean by

"Even with a small gradient in velocity or other transport variables( K and E ) the equations doesn't solves for even iteration. But when I only have the temporal , advection, diffusion and dissipation term, with solver converges with uniform initial conditions all over the domain."

Could you give more details? Thanks!

**Reply by SoorajBharan** [reply]
Hi @richardschu 
Thank you for your suggestions. I tried using the BDFCoupledScheme, but unfortunately, the solver still did not converge.

Here is the approach I followed to build the solver:
I took the incompressible_flow_with_transport solver and modified it to take two scalar equations for $\mathcal{K}$ and $\mathcal{E}$ similar to temperature. The driver.cpp to enable communication between convection_diffusion solver for $\mathcal{K}$ and convection_diffusion solver for $\mathcal{E}$.
Based on the values of $\mathcal{K}$ and $\mathcal{E}$ I updated the scalar viscosity. 
Then I build a channel flow test case for these equations. The type of boundary conditions are given in previous images.
I gave reynolds number, channel half width and kinematic viscosity are input parameters. Based on these values I calculated $U_{centerline}$. $\mathcal{K}$ and $\mathcal{E}$ are then calculated based on $U_{centerline}$
Example:
Re = 1000;
half width($\delta$) = 0.5;
kinematic viscosity($\nu$) = 1.0e-6(water) $m^{2}/s$;
$U_{centerline} = 0.002 m/s
k = 1.5e-8 $m^{2}/s^{2}$
$\mathcal{K}$ = -18.015
$\epsilon$ = 4.312e-12 $m^{2}/s^{3}$
$\mathcal{E}$ = -26.169
$\nu_T$ = 4.696e-6
$\nu_T / \nu$ = 4.696
I used these value at the inlet of the channel.
With these boundary conditions I was not able to solve the above mentioned plain scalar transport equations. 

Later I adjusted the viscosity to get positive values for $\mathcal{K}$ and $\mathcal{E}$ at the inlet. I was able to obtain some solution at  the $\nu$ = 10e-2.
$\mathcal{K}$ = 0.405
$\mathcal{E}$ = 1.461
$\nu_T$ = 0.04696
$\nu_T / \nu$ = 4.696

Once the solver was running for these boundary condition I added the production and dissipation terms as volume integrals using a RHSKernel (I made a copy of rhs_operator.h and rhs_operator.cpp to /convection_diffusion/spatial_discretization/operators/). And added the cross derivative terms for variable viscosity and positivity limiters to the diffusion operators as volume integrals.

With these modifications I was not able to solve the channel flow case.

Thank You

**Comment by nfehn** [comment]
Hi @SoorajBharan, thanks for starting these discussion in addition to the email exchange we had previously. It's great to see that you address these topics and challenges with ExaDG.

Right now it's hard to say where your problems come from. I don't think that the handling of the variable viscosity is a major issue that causes these problems, as long as you access valid data and not some wrong numbers (including e.g. negative values).

I do not fully understand the equations you provide. In the transport equations, the two terms in the end with $\nu_T/\epsilon$ would cancel each other with the minus sign as posted originally!? In the other equation ($K$ equation), the last two terms could be added up to $3 \nu_T/\epsilon$?

I suggest the following procedure: (i) We need to be sure that the mathematical model is correct, (ii) we need to be sure that the mathematical model is implemented correctly according to the syntax and interfaces of ExaDG/dealii and we need to discuss which discrete DG formulations are used for these terms, (iii) we need to figure out whether certain discrete formulations are problematic in terms of stability w.r.t. the formulation realized in (i) and (ii). Regarding aspect (iii), you could take a look at chapter 3 of my PhD thesis (https://mediatum.ub.tum.de/1601025), where stability issues are e.g. addressed for flow-transport coupling. This could be helpful to understand that translating a mathematical model into a discrete model (with spatial discretization) might potentially require special care. It do not want to say that the problem is the same here, but rather point out that we have to be careful in general and can not always expect robust behavior.

**Reply by SoorajBharan** [reply]
Hi [@nfehn],
Thank you for your detailed response and for sharing the link to your thesis.
To provide more clarity, I’ve uploaded the transport equations in the form commonly found in most textbooks:
![k_epsilon_equations_vector_form](https://github.com/user-attachments/assets/5cdf85ec-5705-4344-b2ba-bc7395b49d0d)

In the earlier equations I shared, I had split the diffusion term using the product rule and made simplifications based on the equation for eddy viscosity. I now realize that this may have caused some confusion or introduced errors,

Regarding your suggested procedure:

-     I will carefully recheck the mathematical model to ensure its correctness and consistency.
-     I’ll verify that the implementation aligns with the syntax and interfaces provided by ExaDG and deal.II. To aid clarity, I will upload images of the discrete DG formulation I’ve used for each term.
-     Lastly, I’ll focus on identifying whether specific discrete formulations might be causing stability issues.
I’ll keep working through these steps and post my updated if I am able to solve the issues.

Thank You
